CaloCluster/src/CaloClusterFast_module.cc

```diff
@@ -1,322 +1,197 @@
-           int    roId  = caloDigi.SiPMID();
-           int    crId  = roId / nro; //cal->caloIDMapper().crystalIDFromSiPMID(roId); use short form for efficiency
-           //double adc2MeV  = calorimeterCalibrations->ADC2MeV(roId);
-           const auto& waveform = caloDigi.waveform();
-
-           auto it   = waveform.begin();
-           auto tail = waveform.begin();
-           unsigned countdown(window_ - 1);
-           unsigned nCount(0);
-           deque_.clear();
-	  
-           for (; it != waveform.end(); ++it)
-           {
-                while (!deque_.empty() && *it > deque_.back()) deque_.pop_back();
-                deque_.push_back(*it);
-
-                if (countdown > 0) countdown--;
-                else
-                {
-                   if (deque_.front()> minAmp_ && deque_.front()== *std::prev(it,windowPeak_) && *std::prev(it,windowPeak_) != *std::prev(it,windowPeak_-1))
-                   {
-                       int index = int(t0/digiSampling_) + nCount - winOffsetT0_;
-		      
-                       hitList_[index].push_back(FastHit(crId,index,deque_.front()));
-                       if (deque_.front()> minSeedAmp_) seeds_.emplace_back(&(hitList_[index].back()));
-                   }
-
-                   if (*tail == deque_.front()) deque_.pop_front();
-                   ++tail;
-                }
-                ++nCount;
-           } 
+  //----------------------------------------------------------------------------------------------------------
+  void CaloClusterFast::makeClusters(CaloClusterCollection& caloClusters, const art::Handle<CaloHitCollection>& caloHitsHandle)
+  {
+      const Calorimeter& cal = *(GeomHandle<Calorimeter>());
+      const CaloHitCollection& caloHits(*caloHitsHandle);
+      if (caloHits.empty()) return;
+
+      std::vector<int> hits(caloHits.size());
+      std::iota(hits.begin(),hits.end(),0);
+      
+      auto functorTime = [&caloHits](int a, int b) {return caloHits[a].time() < caloHits[b].time();};
+      std::sort(hits.begin(),hits.end(),functorTime);
+      for (auto val : hits) if (caloHits[val].energyDep() < EnoiseCut_) val = -1;
+
+      auto iterSeed = hits.begin();
+      while (iterSeed != hits.end())
+      {
+          //find the first hit above the energy threshold, and the last hit within the required time window
+          const CaloHit& hitSeed = caloHits[*iterSeed];
+          if (*iterSeed==-1 || hitSeed.energyDep()< EminSeed_) {++iterSeed; continue;}
+          double timeStart = hitSeed.time();
+
+          //find the range around the seed time to search for other hits to form clusters
+          auto iterStart(iterSeed), iterStop(iterSeed);
+          while (iterStop  != hits.end()   && (*iterStop==-1  || caloHits[*iterStop].time() - timeStart < deltaTime_))  ++iterStop;
+          while (iterStart != hits.begin() && (*iterStart==-1 || timeStart - caloHits[*iterStart].time() < deltaTime_)) --iterStart;
```

Comment:
why not use rbegin,rend instead of rolling your own reverse iterator?

Reply:
This is actually a bug if hits has exactly 1 entry.

Reply:
Because you scan back from iterSeed (not the beginning or end of the vector) until you either reach the beginning of the vector or the time difference with the seed hit is too large. This define the lower boundary to include crystal hits in the clustering algorithm (obviously same argument for iterStop going forward). 

if you have a single entry, then iterStart = hits.begin() and the loop exits right away so you are safe.

Reply:
Just checked, it works with a single entry.